core/src/test/java/io/grpc/internal/ServerImplTest.java

```diff
@@ -362,6 +371,120 @@ public void run() {
     server.shutdown();
   }
 
+  @Test
+  public void testCallContextIsBoundInListenerCallbacks() throws Exception {
+    registry.addService(ServerServiceDefinition.builder("Waiter")
+        .addMethod(
+            MethodDescriptor.create(
+                MethodType.UNKNOWN, "Waiter/serve", STRING_MARSHALLER, INTEGER_MARSHALLER),
+            new ServerCallHandler<String, Integer>() {
+              @Override
+              public ServerCall.Listener<String> startCall(
+                  MethodDescriptor<String, Integer> method,
+                  ServerCall<Integer> call,
+                  Metadata headers) {
+                // Record the context present as call start, should remain the same for
+                // all the callbacks.
+                final Context initial = Context.current();
+                assertNotSame(Context.ROOT, initial);
```

Comment:
Discussed offline. Plumbing in a way to use a different Context than ROOT as the base context for all server calls. We can then have a key only in this test and verify that `initial` here contains a value for that key.